### ISSUE TITLE
Wishlist item: maintaining sparse matrices when using SparseMatrixSimilarity

### ISSUE THREAD

**Issue by dschallis** [RaRe-Technologies/gensim#289]
When using sparse data, SparseMatrixSimilarity performs incredibly well.  When it returns query results though, it converts them back into either a dense matrix (numpy.ndarray), or list of tuples.

It'd be great to be able to configure it to just return a sparse matrix instead of a dense one, since it's already working with that structure internally.

E.g. SparseMatrixSimilarity contains the following code, where the 'result' variable is already in a sparse matrix format, and is converted to a dense one before being returned:

```
    # compute cosine similarity against every other document in the collection
    result = self.index * query.tocsc() # N x T * T x C = N x C
    if result.shape[1] == 1 and not is_corpus:
        # for queries of one document, return a 1d array
        result = result.toarray().flatten()
    else:
        # otherwise, return a 2d matrix (#queries x #index)
        result = result.toarray().T
    return result
```

just returning (perhaps if some option passed to the constructor is set):

```
result = result.transpose()
```

would be very useful. I've currently got some code which calculates similarities on ~25k documents using something like:

```
matrix_sim = SparseMatrixSimilarity(corpus, num_features=len(dictionary))
similarities = matrix_sim[corpus]
return scipy.sparse.csc_matrix(similarities)
```

The above currently takes ~45s, with ~35s of that time being the conversion from dense to sparse matrix.

Returning the sparse matrix directly (using `result = result.transpose()`) reduces the above to taking ~5s (some time is also saved on converting the sparse matrix to dense by gensim).

No problem if this isn't generally useful though, I can always create a new subclass of SparseMatrixSimilarity locally :)

**Comment by piskvorky** [comment]
Thanks for the suggestion, I can see that can be useful.

My worry is that the other interfaces in gensim won't be expecting such CSC matrix, so there may be "mismatch errors" if you try to use the scipy.sparse result in unexpected ways.

But, if the default behaviour doesn't change (=this new functionality is an I-know-what-I'm-doing optional switch), then a pull request will be welcome!

**Comment by dschallis** [comment]
Great, I'll do some more testing with this, and submit a pull request if it goes well.

It's always annoying having separate code for sparse/nparrays, I always wished there was some numpy/scipy class which abstracted over all of them :)

**Comment by piskvorky** [comment]
Hello @dschallis -- any progress on this?

**Comment by davechallis** [comment]
@piskvorky Oops, had forgotten about this... I think I've got a working implementation somewhere around, will dig it out next week and hopefully add a pull request for it (I think it was a pretty easy change anyway).

**Comment by menshikh-iv** [comment]
Resolved in #590 and another PRs.
